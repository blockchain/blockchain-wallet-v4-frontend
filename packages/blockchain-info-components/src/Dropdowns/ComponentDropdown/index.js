import React from 'react'
import PropTypes from 'prop-types'
import onClickOutside from 'react-onclickoutside'

import Dropdown from './template'

class ComponentDropdown extends React.PureComponent {
  constructor (props) {
    super(props)

    this.state = {
      toggled: props.opened
    }
    this.handleClick = this.handleClick.bind(this)
    this.handleCallback = this.handleCallback.bind(this)
  }

  handleClick () {
    this.setState({ toggled: !this.state.toggled })
  }

  handleClickOutside () {
    this.setState({ toggled: false })
  }

  handleCallback (item) {
    const { toggled } = this.state
    const { toggleOnCallback } = this.props
    this.setState({
      toggled: toggleOnCallback ? false : toggled,
      selectedComponent: this.props.forceSelected
        ? this.props.selectedComponent
        : item
    })
    if (this.props.callback) {
      this.props.callback(item)
    }
  }

  render () {
    const { ...rest } = this.props

    return (
      <Dropdown
        {...rest}
        handleClick={this.handleClick}
        handleCallback={this.handleCallback}
        toggled={this.state.toggled}
        selectedComponent={this.props.selectedComponent}
      />
    )
  }
}

ComponentDropdown.defaultProps = {
  color: 'brand-secondary',
  toggleOnCallback: true,
  uppercase: false,
  opened: false,
  down: false
}

ComponentDropdown.propTypes = {
  callback: PropTypes.func.isRequired,
  toggleOnCallback: PropTypes.bool,
  opened: PropTypes.bool,
  uppercase: PropTypes.bool,
  down: PropTypes.bool
}

export default onClickOutside(ComponentDropdown)
