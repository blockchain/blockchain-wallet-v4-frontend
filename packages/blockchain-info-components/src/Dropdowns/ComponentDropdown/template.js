import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

import { Icon } from '../../Icons'
import { Palette } from '../../Colors'
import { keysIn } from 'ramda'

const Wrapper = styled.div`
  display: inline-flex;
  text-transform: ${props => (props.uppercase ? 'uppercase' : 'none')};
  position: relative;
`
const ButtonContainer = styled.div`
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  width: inherit;
`
const Button = styled.div`
  display: inline;
`
const DropdownIcon = styled(Icon)`
  cursor: pointer;
  padding-left: 4px;
`
const DropdownList = styled.ul`
  background-clip: padding-box;
  background-color: ${props => props.theme['white']};
  border: 1px solid ${props => props.theme['gray-1']};
  border-radius: 4px;
  bottom: 0px;
  box-sizing: border-box;
  box-shadow: 0 6px 12px rgba(0, 0, 0, 0.175);
  display: ${props => (props.toggled ? 'block' : 'none')};
  float: none;
  height: auto;
  width: inherit;
  line-height: 20px;
  list-style-image: none;
  list-style-position: outside;
  list-style-type: none;
  margin: 2px 0px;
  min-width: 20px;
  overflow: auto;
  padding: 5px;
  position: absolute;
  right: 0;
  ${props =>
    props.down
      ? 'top: 25px; bottom: auto;'
      : 'top: auto; bottom: 25px;'} z-index: 10;
`

const DropdownItem = styled.li`
  color: ${props => props.theme['gray-5']};
  cursor: pointer;
  font-family: 'Montserrat', Helvetica, sans-serif;
  font-size: 14px;
  font-weight: 300;
  text-align: left;
  text-size-adjust: 100%;
  white-space: nowrap;
`

const Dropdown = props => {
  const {
    color,
    down,
    uppercase,
    toggled,
    selectedComponent,
    components,
    handleClick,
    handleCallback
  } = props

  return (
    <Wrapper uppercase={uppercase}>
      <DropdownList toggled={toggled} down={down} data-e2e={props['data-e2e']}>
        {components.map((comp, index) => {
          return (
            <DropdownItem key={index} onClick={handleCallback.bind(null, comp)}>
              {comp}
            </DropdownItem>
          )
        })}
      </DropdownList>
      <ButtonContainer color={color} onClick={handleClick}>
        <Button>{selectedComponent}</Button>
        <DropdownIcon
          name='down-arrow'
          size='12px'
          data-e2e='dropdownToggleButton'
        />
      </ButtonContainer>
    </Wrapper>
  )
}

Dropdown.defaultProps = {
  color: 'brand-secondary',
  toggled: false,
  selectedValue: 0,
  uppercase: true,
  down: false
}

Dropdown.propTypes = {
  selectedValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  callback: PropTypes.func.isRequired,
  toggled: PropTypes.bool,
  color: PropTypes.oneOf(keysIn(Palette())),
  uppercase: PropTypes.bool,
  down: PropTypes.bool
}

export default Dropdown
