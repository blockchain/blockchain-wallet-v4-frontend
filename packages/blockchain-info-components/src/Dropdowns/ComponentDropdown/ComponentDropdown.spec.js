import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import ComponentDropdown from './template'

describe('ComponentDropdown component', () => {
  it('default renders correctly', () => {
    const components = [{ text: 'test', value: 'value' }]
    const component = shallow(
      <ComponentDropdown
        components={components}
        callback={() => {}}
        handleClick={() => {}}
        handleCallback={() => {}}
      />
    )
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
})
