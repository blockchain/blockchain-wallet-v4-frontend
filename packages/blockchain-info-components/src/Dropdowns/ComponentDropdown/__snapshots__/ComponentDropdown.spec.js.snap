// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentDropdown component default renders correctly 1`] = `
<template__Wrapper
  uppercase={true}
>
  <template__DropdownList
    down={false}
    toggled={false}
  >
    <template__DropdownItem
      key="0"
      onClick={[Function]}
    >
      <Component />
    </template__DropdownItem>
  </template__DropdownList>
  <template__ButtonContainer
    color="brand-secondary"
    onClick={[Function]}
  >
    <template__Button />
    <template__DropdownIcon
      data-e2e="dropdownToggleButton"
      name="down-arrow"
      size="12px"
    />
  </template__ButtonContainer>
</template__Wrapper>
`;
